### ISSUE TITLE
26.9.3: concurrent sampled requests on dense Qwen3.8-27B fail with "Cannot reshape array of size N into shape (k,248320)" in the grouped MTP verify (bisected to #434; MLX_SERVE_MTP_BATCH_CORR=0 avoids it)

### ISSUE THREAD

**Issue by rotemben-droid** [ddalcu/mlx-serve#446]
## Summary

On 26.9.3, whenever two or more requests on a dense Qwen3.8-27B 4-bit model share one MTP verify round **and every request in the group is sampled (temperature > 0)**, the round aborts and every request in the group returns HTTP 500 `generation failed`:

```
[mlx] [reshape] Cannot reshape array of size 744960 into shape (2,248320). at .../lib/mlxc-src/mlx/c/ops.cpp:2712
[scheduler] batched verify aborted: MLX failure mid-generation (MlxFailure) — failing these requests, the server keeps serving
  -> GenerationFailed
  -> GenerationFailed
```

The server survives. Single requests are fine. Greedy groups are fine. A mixed group (one greedy + one sampled) is fine. Reproduces 100% on a freshly started server on two different Macs.

Bisection by knob (details below) isolates it to the sampled group-accept on the joined verify block, `Generator.mtpGroupSampledAccept` / `groupProbsBlock` in `src/generate.zig`, added in **62c4f8d `perf(mtp): one filtered verify block per group, one filter pass, a masked greedy verify (#434)`**, committed ~4 h before the v26.9.3 tag. `MLX_SERVE_MTP_BATCH_CORR=0` (the kill switch that function checks first) makes the identical concurrent load pass with the batched verify still engaged and no throughput loss.

The reshape sizes are always a multiple of the vocab (248320) but never equal to the group size `k`: 744960 = 3 rows → (2, V); 1241600 = 5 rows → (2, V); 1986560 = 8 rows → (3, V) and → (4, V); 1489920 = 6 rows → (4, V). I.e. something in the sampled accept path treats the group's joined `[1, Σ(1+m_i), V]` block as having one row per slot. It only triggers when `use_joined` is true (`k == gens.len`, every row sampled), which is exactly the greedy/mixed vs all-sampled split observed.

## Environment

- mlx-serve 26.9.3 release tarball (`--version`: mlx-serve 26.9.3, mlx 0.32.2, mlx-c 56b2d39fc831, ggml 0.23.0 (5266f24da), llama.cpp b10809), macOS 26.6.2
- Mac A: M4 Max 48 GB, `chimingw/Qwen3.8-27B-Uncensored-OrcaRouter-MLX-4bit` (nax off), `--kv-quant 8 --max-concurrent 2`
- Mac B: M5 Max 48 GB, `ddalcu/Qwen3.8-27B-MLX-Serve-4bit` (nax on), kv-quant off, `--max-concurrent 1`
- Both packs load the native in-checkpoint MTP head (`[mtp] loaded native MTP head (dense-mlp; …)`, depth 6 on A, depth 8 / profile `g17_nax_q4_gs64` on B). Log reports the trunk as `Model: qwen3_5_moe (64 layers, 5120-dim, head_dim=256, 24h/4kv, 4-bit affine quant)`.
- A MoE `Qwen3.8-Flash-Next` (M3 Ultra, `--mtp`) on the same 26.9.3 build served ~600 requests in the same window with zero failures.

Full flags (Mac A; Mac B differs only as noted above plus `--reasoning-budget 0 --max-resident-models 2 --skip-mem-preflight`):
```
mlx-serve --model <dir> --serve --host 0.0.0.0 --lan-share all --port 11234 --ctx-size 131072 --kv-quant 8 \
  --reasoning-budget 8192 --prefix-cache-mem 4GB --prefix-cache-disk 64GB --max-concurrent 2
MLX_SERVE_CACHE_LIMIT=6442450944
```

## Reproduction

Fresh server, one warm-up request, then two identical **sampled** requests ~0.3 s apart:

```bash
for i in 1 2; do
  curl -s http://HOST:11234/v1/chat/completions -H 'Content-Type: application/json' -d '{
    "model":"Qwen3.8-27B-MLX-Serve-4bit",
    "messages":[{"role":"user","content":"Write a detailed 400-word essay about the history of lighthouses. Do not think, answer directly."}],
    "max_tokens":200,"temperature":0.7}' &
done; wait
```

Both → `500 {"error":{"message":"generation failed","type":"server_error","code":500}}` within 0.3–0.8 s. Same with three requests. Same at temperature 0.8 and 1.0. The same two requests at `"temperature":0` → both 200. One at 0 + one at 0.7 → both 200.

## Bisection matrix (same machines, same model, same flags, minutes apart; each cell = a fresh start + warm-up + 2 concurrent + 3 concurrent + 1 single after)

| Build / env | 2 concurrent sampled | 3 concurrent sampled | 2 concurrent greedy | aborts in log |
|---|---|---|---|---|
| 26.9.2 (brew), Mac A | 200/200, 12–13 tok/s each | — | — | 0 |
| 26.9.2 (brew), Mac B | 200/200, 25 tok/s each | 200×3, 13 tok/s each | — | 0 |
| **26.9.3**, Mac A | **500/500** | — | 200/200 | 2 |
| **26.9.3**, Mac B | **500/500** | **500/500** + 200 (third slot solo) | 200/200 | 4 (`(3,V)`, `(4,V)`) |
| 26.9.3 + `MLX_SERVE_MTP_BATCHED=0`, Mac A | 200/200, 10 tok/s each | 200×3, 6 tok/s each | — | 0 |
| 26.9.3 + `MLX_SERVE_MTP_BATCHED=0`, Mac B | 200/200, 20 tok/s each | 200×3, 10 tok/s each | — | 0 |
| **26.9.3 + `MLX_SERVE_MTP_BATCH_CORR=0`, Mac B** | **200/200, 39 tok/s each** | **200×3, 33 tok/s each** | 200/200, 43 tok/s | 0, `[batched] gdn batched verify engaged (slots=2, width=8)` |

So: the grouped verify itself (fd9c0c3, `runBatchedMtpTickInner` → `forwardMoeBatchedVerify`) works and is a real win (39 vs 20 tok/s per stream); the abort comes from the sampled accept over the joined block that #434 layered on top. `MLX_SERVE_MTP_BATCH_CORR=0` is a workable mitigation on 26.9.3.

## Log excerpt (Mac A, 26.9.3, the reproduction above)

```
POST /v1/chat/completions (1 msgs, max_tokens=200, temp=0.70, top_p=0.95, top_k=20, stream=false, thinking=false, sys=0b, user=96b, tools=0b, tool_msgs=0)
  prompt=36 tokens, max_gen=200, ctx=131072
  spec-gate: ngram-score=0.000 (threshold=0.010)
  pld=disabled (ngram-score=0.000 < gate threshold 0.010)
  [hot-cache] reused 5/36 tokens (matched 36; entry 23/23)
  pld=enabled (draft_len=5, key_len=3)
POST /v1/chat/completions (1 msgs, max_tokens=200, temp=0.70, top_p=0.95, top_k=20, stream=false, thinking=false, sys=0b, user=96b, tools=0b, tool_msgs=0)
  prompt=36 tokens, max_gen=200, ctx=131072
  spec-gate: ngram-score=0.000 (threshold=0.010)
  pld=disabled (ngram-score=0.000 < gate threshold 0.010)
  [hot-cache] reused 5/36 tokens (matched 36; entry 23/23)
[mlx] [reshape] Cannot reshape array of size 744960 into shape (2,248320). at /Users/runner/work/mlx-serve/mlx-serve/lib/mlxc-src/mlx/c/ops.cpp:2712
[scheduler] batched verify aborted: MLX failure mid-generation (MlxFailure) — failing these requests, the server keeps serving
  -> GenerationFailed
  -> GenerationFailed
```

Under real traffic on Mac B (a ~1275-token single-image vision request at temp 1.0 grouped with a long streaming tool-call session at temp 1.0) it hit six times in 3.5 h, e.g.:

```
  Inserted 1024 image + 0 video + 0 audio soft tokens at position 3 (prompt: 250 -> 1276 tokens)
  prompt=1276 tokens, max_gen=1200, ctx=131072
  spec-gate: ngram-score=0.016 (threshold=0.010)
[mlx] [reshape] Cannot reshape array of size 1241600 into shape (2,248320). at /Users/runner/work/mlx-serve/mlx-serve/lib/mlxc-src/mlx/c/ops.cpp:2712
  -> GenerationFailed
[scheduler] batched verify aborted: MLX failure mid-generation (MlxFailure) — failing these requests, the server keeps serving
  -> streaming error: error.GenerationFailed
```

## Two smaller observations

- Mac B runs `--max-concurrent 1`, yet `/props` shows `"total_slots":1` with `"batching":{"supported":true,"max_group":32}` and the log shows 2–4-row verify groups. If `--max-concurrent 1` is meant to serialise, it no longer does for the MTP group.
- Because the abort takes the whole group, one short request can kill an unrelated 65k-token streaming session that happened to share the round. A per-row failure (or falling back to per-row accept when the joined block shape does not match) would contain the blast radius even before the root cause is fixed.

## Expected

All-sampled groups verify like greedy/mixed groups do, or the sampled accept declines the joined block and builds per-row, as the doc comment on `mtpGroupSampledAccept` says it does when a row is padded.

Happy to test a build or run any extra env knob / trace; full `~/.mlx-serve/logs/mlx-serve-11234.log` from either Mac available on request.

**Comment by ddalcu** [comment]
🤖 **MLXServe-PRBot**

Your bisection lands on real code, and I can point at the exact line the traceback comes from. I reviewed and approved #434 and did not catch this, so taking the correction: I checked that the `offs` walk and the projection's `offset += row.seq` walk agree when `k == gens.len`, and treated the `getShape(joined)[1] == total` check as sufficient to make the joined block safe. It is not, because of what happens one layer down.

**The reshape.** `Generator.mtpBatchedAcceptGraph` (`src/generate.zig:6078`) does:

```zig
const shape = mlx.getShape(probs_all); // [1, 1+m, V]
const vocab = shape[2];
const p2_shape = [_]c_int{ mi + 1, vocab };
try mlx.check(mlx.mlx_reshape(&p2d, probs_all, &p2_shape, 2, s));
```

It reads `shape[2]` for the vocab and then reshapes to `[m+1, V]` from `m` alone — `shape[1]` is never consulted, and the `// [1, 1+m, V]` comment is an assumption the function does not verify. `mtpBatchedLossyGraph` has the identical pair at `:6236`, so whichever `AcceptRoute` is wired at `:2404` behaves the same way. That is exactly your `(k, V)` target with a source that is a different multiple of the vocab.

**Why the solo path never sees it.** `mtpRoundFinish` hands it `probsAllPositions(verify_logits, …)` where `verify_logits` *is* that row's own `[1, 1+m, V]`, so the assumption is structural. The group path builds the argument instead: `groupProbsRow(probs, offs[j], st.verify_len)` at `:7284`, sliced out of one block. The guard at `:7248` (`st.verify_len != 1 + st.chain.m` → decline the whole group) is what is supposed to keep those equal — and your sizes say they are not equal at the call on this arch, since a 3-position slice being reshaped to 2 rows means `verify_len` and `1 + chain.m` disagreed by one for that row.

**Why greedy and mixed groups escape.** Greedy rows never enter the loop (`temperature <= 0.01` continues), and a mixed group fails `k == gens.len`, so `use_joined` is false and the fallback concatenates the per-row `verify_logits` instead. That matches your matrix exactly, and it is also why `MLX_SERVE_MTP_BATCH_CORR=0` is a clean mitigation — it returns before any of this.

**The part worth fixing regardless of root cause.** That reshape is unchecked, so any disagreement becomes an MLX raise that fails every request in the group rather than declining to the per-row accept. `mtpBatchedAcceptGraph` deriving its row count from `shape[1]`, or refusing when `shape[1] != m + 1`, turns this class from a 500 into a fallback — the same shape as the named `.kv_floor` / `.geometry` declines elsewhere in that file. Worth having even once the underlying mismatch is found, because it is the difference between a degraded round and two dead requests.

**Comment by ddalcu** [comment]
Peak AI !   My PRBot is taking responsibility for its actions 😆 

Thanks for your report @rotemben-droid will have a fix for this soon !

**Comment by rotemben-droid** [comment]
My first ever bug report to mlx-serve, i'm humbled and excited. Thank you so much for all your great work. feel free to ping me for help with testing if you need, heavy user of all your models, on multiple less powerfull machines.

**Comment by ddalcu** [comment]
fixed in latest release !  Thank you
